From 3663d9bf23a9ce93dd22a1cd958c2df44d245598 Mon Sep 17 00:00:00 2001
From: Amithash Prasad <amithash@meta.com>
Date: Thu, 11 Sep 2025 13:47:05 -0700
Subject: [PATCH] pldmtool: Use a non-listening transport for command access

pldmtool does not really need to create a listening (bind) transport
to do a simple send/receive transaction. Hence refactor PldmTransport
allowing a user to specify if the required transport is expected
to be listening or non-listening and switch pldmtool to use a
non-listening transport.

Change-Id: I390298397d71c0ea27234880b844898c7a100ca5
Signed-off-by: Amithash Prasad <amithash@meta.com>
---
 common/transport.cpp         | 16 +++++++++-------
 common/transport.hpp         |  2 +-
 pldmtool/pldm_cmd_helper.cpp |  2 +-
 3 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/common/transport.cpp b/common/transport.cpp
index 5c9678af991..f8bcdc3aadb 100644
--- a/common/transport.cpp
+++ b/common/transport.cpp
@@ -8,7 +8,8 @@
 #include <ranges>
 #include <system_error>
 
-struct pldm_transport* transport_impl_init(TransportImpl& impl, pollfd& pollfd);
+struct pldm_transport* transport_impl_init(TransportImpl& impl, pollfd& pollfd,
+                                           bool listening);
 void transport_impl_destroy(TransportImpl& impl);
 
 static constexpr uint8_t MCTP_EID_VALID_MIN = 8;
@@ -64,7 +65,7 @@ static constexpr uint8_t MCTP_EID_VALID_MAX = 255;
 }
 
 [[maybe_unused]] static struct pldm_transport* pldm_transport_impl_af_mctp_init(
-    TransportImpl& impl, pollfd& pollfd)
+    TransportImpl& impl, pollfd& pollfd, bool listening)
 {
     impl.af_mctp = nullptr;
     pldm_transport_af_mctp_init(&impl.af_mctp);
@@ -85,7 +86,7 @@ static constexpr uint8_t MCTP_EID_VALID_MAX = 255;
     }
 
     /* Listen for requests on any interface */
-    if (pldm_transport_af_mctp_bind(impl.af_mctp, nullptr, 0))
+    if (listening && pldm_transport_af_mctp_bind(impl.af_mctp, nullptr, 0))
     {
         return nullptr;
     }
@@ -100,12 +101,13 @@ static constexpr uint8_t MCTP_EID_VALID_MAX = 255;
     return pldmTransport;
 }
 
-struct pldm_transport* transport_impl_init(TransportImpl& impl, pollfd& pollfd)
+struct pldm_transport* transport_impl_init(TransportImpl& impl, pollfd& pollfd,
+                                           [[maybe_unused]] bool listening)
 {
 #if defined(PLDM_TRANSPORT_WITH_MCTP_DEMUX)
     return pldm_transport_impl_mctp_demux_init(impl, pollfd);
 #elif defined(PLDM_TRANSPORT_WITH_AF_MCTP)
-    return pldm_transport_impl_af_mctp_init(impl, pollfd);
+    return pldm_transport_impl_af_mctp_init(impl, pollfd, listening);
 #else
     return nullptr;
 #endif
@@ -120,9 +122,9 @@ void transport_impl_destroy(TransportImpl& impl)
 #endif
 }
 
-PldmTransport::PldmTransport()
+PldmTransport::PldmTransport(bool listening)
 {
-    transport = transport_impl_init(impl, pfd);
+    transport = transport_impl_init(impl, pfd, listening);
     if (!transport)
     {
         throw std::system_error(ENOMEM, std::generic_category());
diff --git a/common/transport.hpp b/common/transport.hpp
index e9e1808ec28..c7a1ec0274e 100644
--- a/common/transport.hpp
+++ b/common/transport.hpp
@@ -19,7 +19,7 @@ union TransportImpl
 class PldmTransport
 {
   public:
-    PldmTransport();
+    PldmTransport(bool listening = true);
     PldmTransport(const PldmTransport& other) = delete;
     PldmTransport(const PldmTransport&& other) = delete;
     PldmTransport& operator=(const PldmTransport& other) = delete;
diff --git a/pldmtool/pldm_cmd_helper.cpp b/pldmtool/pldm_cmd_helper.cpp
index e4885574924..ae74d911e6c 100644
--- a/pldmtool/pldm_cmd_helper.cpp
+++ b/pldmtool/pldm_cmd_helper.cpp
@@ -130,7 +130,7 @@ int CommandInterface::pldmSendRecv(std::vector<uint8_t>& requestMsg,
     }
 
     auto tid = mctp_eid;
-    PldmTransport pldmTransport{};
+    PldmTransport pldmTransport(false);
     uint8_t retry = 0;
     int rc = PLDM_ERROR;
 
-- 
2.47.3

